File: Factory_Pattern/afc.cpp
#include <algorithm>
#include <iostream>
#include <memory>
#include <vector>

class Dough {};
class Sauce {};
class Cheese {};
class Pepperoni {};
class Clam {};
class Veggie {};

// abstract factory class
class PizzaIngredientFactory {
   public:
    virtual std::unique_ptr<Dough> createDough() = 0;
    virtual std::unique_ptr<Sauce> createSauce() = 0;
    virtual std::unique_ptr<Cheese> createCheese() = 0;
    virtual std::unique_ptr<Pepperoni> createPepperoni() = 0;
    virtual std::unique_ptr<Clam> createClam() = 0;
    // virtual std::vector<std::unique_ptr<Veggie>> createVeggies() = 0;
};

class ThinCrustDough : public Dough {};
class MarinaraSauce : public Sauce {};
class ReggianoCheese : public Cheese {};
class SlicedPepperoni : public Pepperoni {};
class FreshClam : public Clam {};
class Garlic : public Veggie {};
class Onion : public Veggie {};
class Mushroom : public Veggie {};
class RedPepper : public Veggie {};

class NYPizzaIngredientFactory : public PizzaIngredientFactory {
   public:
    std::unique_ptr<Dough> createDough() override { return std::make_unique<ThinCrustDough>(); }
    std::unique_ptr<Sauce> createSauce() override { return std::make_unique<MarinaraSauce>(); }
    std::unique_ptr<Cheese> createCheese() override { return std::make_unique<ReggianoCheese>(); }
    std::unique_ptr<Pepperoni> createPepperoni() override { return std::make_unique<SlicedPepperoni>(); }
    std::unique_ptr<Clam> createClam() override { return std::make_unique<FreshClam>(); }

    // std::vector<std::unique_ptr<Veggie>> createVeggies() {
    //     return {
    //         std::make_unique<Garlic>(),
    //         std::make_unique<Onion>(),
    //         std::make_unique<Mushroom>(),
    //         std::make_unique<RedPepper>()};
    // }
};

class ThickCrustDough : public Dough {};
class PlumTomatoSauce : public Sauce {};
class Mozzarella : public Cheese {};
class FrozenClam : public Clam {};
class Eggplant : public Veggie {};
class Spinach : public Veggie {};
class BalckOlives : public Veggie {};

class ChicagoPizzaIngredientFactory : public PizzaIngredientFactory {
   public:
    std::unique_ptr<Dough> createDough() override { return std::make_unique<ThickCrustDough>(); }
    std::unique_ptr<Sauce> createSauce() override { return std::make_unique<PlumTomatoSauce>(); }
    std::unique_ptr<Cheese> createCheese() override { return std::make_unique<Mozzarella>(); }
    std::unique_ptr<Pepperoni> createPepperoni() override { return std::make_unique<SlicedPepperoni>(); }
    std::unique_ptr<Clam> createClam() override { return std::make_unique<FrozenClam>(); }

    // std::vector<std::unique_ptr<Veggie>> createVeggies() {
    //     return {
    //         std::make_unique<Eggplant>(),
    //         std::make_unique<Spinach>(),
    //         std::make_unique<BalckOlives>()};
    // }
};

class Pizza {
   protected:
    std::string name;
    std::unique_ptr<Dough> dough;
    std::unique_ptr<Sauce> sauce;
    std::unique_ptr<Cheese> cheese;
    std::unique_ptr<Pepperoni> pepperoni;
    std::unique_ptr<Clam> clam;
    // std::vector<std::unique_ptr<Veggie>> veggies;

   public:
    virtual void prepare() = 0;

    void bake() { std::cout << "Bake for 25 min at 350\n"; }
    void cut() { std::cout << "Cutting the pizza into diagonal slices\n"; }
    void box() { std::cout << "Place pizza in official PizzaStore box\n"; }

    void setName(const std::string& name) { this->name; }
    std::string getName() { return this->name; }
};

class CheezePizza : public Pizza {
    std::unique_ptr<PizzaIngredientFactory> ingredientFactory;

   public:
    explicit CheezePizza(std::unique_ptr<PizzaIngredientFactory> iF)
        : ingredientFactory(std::move(iF)) {}

    void prepare() override {
        std::cout << "Preparing " << this->getName() << "\n";
        this->dough = std::move(this->ingredientFactory->createDough());
        this->sauce = std::move(this->ingredientFactory->createSauce());
        this->cheese = std::move(this->ingredientFactory->createCheese());
    }
};

class ClamPizza : public Pizza {
    std::unique_ptr<PizzaIngredientFactory> ingredientFactory;

   public:
    explicit ClamPizza(std::unique_ptr<PizzaIngredientFactory> iF)
        : ingredientFactory(std::move(iF)) {}

    void prepare() override {
        std::cout << "Preparing " << this->getName() << "\n";
        this->dough = std::move(this->ingredientFactory->createDough());
        this->sauce = std::move(this->ingredientFactory->createSauce());
        this->cheese = std::move(this->ingredientFactory->createCheese());
    }
};

class PizzaStore {
   public:
    std::unique_ptr<Pizza> orderPizza(std::string pizza_type) {
        std::unique_ptr<Pizza> pizza = std::move(this->createPizza(pizza_type));
        pizza->prepare();
        pizza->bake();
        pizza->cut();
        pizza->box();
        return pizza;
    }

    virtual std::unique_ptr<Pizza> createPizza(const std::string& pizzaType) = 0;
};

class NYPizzaStore : public PizzaStore {
   protected:
    std::unique_ptr<Pizza> createPizza(const std::string& item) override {
        std::unique_ptr<Pizza> pizza = nullptr;
        std::unique_ptr<PizzaIngredientFactory> iF = nullptr;

        if (item == "Cheese")
            pizza = std::make_unique<CheezePizza>(std::move(iF));
        else if (item == "Clam")
            pizza = std::make_unique<ClamPizza>(std::move(iF));
        else
            pizza = nullptr;
        return pizza;
    }
};

int main() {
    std::unique_ptr<PizzaStore> nyPizzaStore = std::make_unique<NYPizzaStore>();
    std::unique_ptr<Pizza> nyPizza = nyPizzaStore->orderPizza("Cheese");
    nyPizza->setName("NY Cheese Cheese Pizza");
    std::cout << "Ethan ordered a " << nyPizza->getName() << "\n";
}

File: Factory_Pattern/afc.py
from abc import ABC, abstractmethod

class Dough(ABC): pass
class Sauce(ABC): pass
class Cheese(ABC): pass
class Pepperoni(ABC): pass
class Clam(ABC): pass


class PizzaIngredientFactory(ABC):
    @abstractmethod 
    def create_dough(self) -> Dough: pass

    @abstractmethod
    def create_sauce(self) -> Sauce: pass

    @abstractmethod
    def create_cheese(Self) -> Cheese: pass

    @abstractmethod
    def create_pepperoni(self) -> Pepperoni: pass

    @abstractmethod
    def create_clam(self) -> Clam: pass


class Thin_crust_dough(Dough): pass
class Marinara_sauce(Sauce): pass
class Reggiano_cheese(Cheese): pass
class Sliced_pepperoni(Pepperoni): pass
class Fresh_clam(Clam): pass


class NYPizzaIngredientFactory(PizzaIngredientFactory):
    def create_dough(self) -> Dough: return Thin_crust_dough()
    def create_sauce(self) -> Sauce: return Marinara_sauce()
    def create_cheese(Self) -> Cheese: return Reggiano_cheese()
    def create_pepperoni(self) -> Pepperoni: return Sliced_pepperoni()
    def create_clam(self) -> Clam: return Fresh_clam()


class Pizza(ABC):
    @abstractmethod
    def prepare(self) -> None: pass

    def bake(self) -> None: print("Bake for 25 min at 350")
    def cut(self) -> None: print("Cutting the pizza into diagonal slices")
    def box(self) -> None: print("Place pizza in official PizzaStore box")


class Cheese_pizza(Pizza):
    def __init__(self, IF: PizzaIngredientFactory) -> None:
        self.IF: PizzaIngredientFactory = IF

    def prepare(self) -> None:
        print(f"Prepare {self.pygame.mixer.set_num_channels()}")

        self.dough: Dough = self.IF.create_dough()
        self.sauce: Sauce = self.IF.create_sauce()
        self.cheese: Cheese = self.IF.create_cheese()

class PizzaStore(ABC):
    @abstractmethod
    def create_pizza(self, pizza_type: str) -> Pizza: pass
    
    def order_pizza(self, pizza_type: str) -> Pizza:
        self.pizza: Pizza = self.create_pizza(pizza_type)
        self.pizza.prepare()
        self.pizza.bake()
        self.pizza.cut()
        self.pizza.box()
        return Pizza

class NY_pizza_store(PizzaStore):
    def create_pizza(self, pizza_type: str) -> Pizza:
        if pizza_type == "Cheese":
            return Cheese_pizza()
        else: 
            None

def main() -> None:
    ny_pizza_store = NY_pizza_store()
    ny_pizza_store.order_pizza("Cheese")
    print("Ethan ordered a pizza")


if __name__ == '__main__':
    main()
    